src/types.jl

```diff
@@ -0,0 +1,154 @@
+abstract AbstractODE
+
+
+type ExplicitODE <: AbstractODE
+    t0; y0
+    F! :: Function
+    jac! :: Function
+end
+
+
+function explicit_ineff(t0,y0,F;jac = (t,y)->fdjacobian(F,t,y))
+    function F!(t,y,dy)
+        # this is why we can't handle a scalar type any more
+        dy[:] = F(t,y)
+    end
+    function jac!(t,y,J)
+        J[:] = jac(t,y)
+    end
+    return ExplicitODE(t0,y0,F!,jac!)
+end
+
+
+abstract AbstractStepper
+abstract AbstractState
+
+
+# this might suffice for some solvers
+type Step{T,S}
+    t  :: T
+    y  :: S
+    dy :: S
+end
+
+
+# for debugging
+function show(io::IO, state :: Step)
+    println("t  =$(state.t)")
+    println("y  =$(state.y)")
+    println("dy =$(state.dy)")
+end
+
+
+# general purpose options
+immutable Options{T}
+    # stepper options
+    initstep :: T
+    tstop    :: T
+    reltol   :: T
+    abstol   :: T
+    minstep  :: T
+    maxstep  :: T
+    norm     :: Function
+    maxiters :: T
+
+    # dense output options
+    tspan    :: Vector{T}
+    points   :: Symbol
+    stopevent :: Function
+    roottol  :: T
+
+    function Options(;
```

Comment:
Maybe add the usual positional constructor, add the invariants to that and pipe the keyword constructor through that.  If this is immutable we also need an updating constructor, e.g. `Options(opts, tspan=[1,4])` to make a new options-struct with different tspan.